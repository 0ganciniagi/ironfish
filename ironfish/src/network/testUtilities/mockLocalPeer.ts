/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import ws from 'ws'
import wrtc from 'wrtc'
import { PrivateIdentity } from '../identity'
import { LocalPeer } from '../peers/localPeer'
import { mockPrivateIdentity } from './mockPrivateIdentity'

/**
 * Utility to create a fake "keypair" for testing the network layer
 */
export function mockLocalPeer({
  identity = mockPrivateIdentity('local'),
  version = 'sdk/1/cli',
}: {
  identity?: PrivateIdentity
  version?: string
} = {}): LocalPeer {
  return new LocalPeer(identity, version, ws, wrtc)
}
