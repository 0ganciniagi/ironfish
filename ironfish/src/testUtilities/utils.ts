/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

/**
 * This is only usable in the jasmine runner
 */
export function getCurrentTestPath(): string {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
  const jasmineAny = global.jasmine as any
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  return jasmineAny.testPath as string
}
