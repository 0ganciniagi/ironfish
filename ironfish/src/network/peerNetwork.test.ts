/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

jest.mock('ws')

import ws from 'ws'
import { PeerNetwork, RoutingStyle } from './peerNetwork'
import { getConnectedPeer, mockPrivateIdentity } from './testUtilities'

jest.useFakeTimers()

it('Closes the PeerManager when close is called', () => {
  const peerNetwork = new PeerNetwork(mockPrivateIdentity('local'), 'sdk/1/cli', ws)
  const stopSpy = jest.spyOn(peerNetwork.peerManager, 'stop')
  peerNetwork.stop()
  expect(stopSpy).toBeCalled()
})

it('Registers a handler', () => {
  const peerNetwork = new PeerNetwork(mockPrivateIdentity('local'), 'sdk/1/cli', ws)
  peerNetwork.registerHandler(
    'hello',
    RoutingStyle.gossip,
    (p) => Promise.resolve(p),
    () => {},
  )
  expect(peerNetwork['routingStyles']).toMatchSnapshot()
  peerNetwork.stop()
})

it('ignores a message if validation fails', async () => {
  const peerNetwork = new PeerNetwork(mockPrivateIdentity('local'), 'sdk/1/cli', ws)
  const handlerMock = jest.fn(() => {})
  peerNetwork.registerHandler(
    'hello',
    RoutingStyle.gossip,
    () => Promise.reject(new Error('invalid message')),
    handlerMock,
  )

  const { peer } = getConnectedPeer(peerNetwork.peerManager)
  const message = { type: 'hello', nonce: 'test_handler1', payload: { test: 'Payload' } }
  await peerNetwork['handleMessage'](peer, { peerIdentity: peer.getIdentityOrThrow(), message })
  expect(handlerMock).not.toBeCalled()
  peerNetwork.stop()
})

it('changes isReady when peers connect', () => {
  const peerNetwork = new PeerNetwork(
    mockPrivateIdentity('local'),
    'sdk/1/cli',
    ws,
    undefined,
    {
      minPeersReady: 1,
    },
  )

  expect(peerNetwork.isReady).toBe(false)

  const readyChanged = jest.fn()
  peerNetwork.onIsReadyChanged.on(readyChanged)

  peerNetwork.start()
  expect(peerNetwork.isReady).toBe(false)

  const { peer } = getConnectedPeer(peerNetwork.peerManager)
  expect(peerNetwork.isReady).toBe(true)

  peer.close()
  expect(peerNetwork.isReady).toBe(false)

  peerNetwork.stop()
  expect(peerNetwork.isReady).toBe(false)

  expect(readyChanged).toBeCalledTimes(2)
  expect(readyChanged).toHaveBeenNthCalledWith(1, true)
  expect(readyChanged).toHaveBeenNthCalledWith(2, false)
})
