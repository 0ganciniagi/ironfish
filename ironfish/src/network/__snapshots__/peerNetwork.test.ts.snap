// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registers a handler 1`] = `
Map {
  "hello" => "gossip",
}
`;
